### PR TITLE
New: Allow custom directory structure under Artist folder.

### PR DESCRIPTION
**Pull request by Qstick**
#### Database Migration
No

#### Description
This PR removes `AlbumFolder` from artist table and model. Additionally, it pulls `AlbumFolderFormat` from naming config and allows the user to build the entire artist sub-directory structure from `TrackFormat`.

image

List of changes:
- Remove `AlbumFolder` property from Artist table/model
- Remove `AlbumFolderFormat` from Naming Config
- Cleanup FileNameBuilder
- Purge BasicNamingConfig used from V2 Sonarr
- Add naming token for primary album type and default `Album Type` token to be Secondary album type, or if blank, primary

Naming structures such as this are now allowed:
- `{Album Type}/{Album Title} [{Release Year}]/Disc {medium:0}/{track:00} - {Track Title}`
- `{Album Title} ({Album Disambiguation}) [{Release Year}]/{Medium Format} {medium:0}/{track:00} - {Track Title}`
- `{Quality Title}/{Album Title} [{Release Year}]/{track:00} - {Track Title}`
- etc...

#### Todos
- [ ] Tests for correct path combination from track format
- [ ] Track format validation
- [ ] Get set write time working for sub directories
- [x] Show sample extension when showing sample in UI
- [ ] DB Migration to remove AlbumFolder column from Artists Table
- [ ] Write DB migration to remove AlbumFolderFormat from NamingConfig table
- [ ] Write DB migration to combine AlbumFolderFormat and TrackFormat so users don't lose their current naming setup.
- [ ] Documentation
- [ ] Decide what to do it renaming is disabled.

#### Issues Fixed or Closed by this PR
#96 
#139 
#393 
#350 
*

### CODE REVIEW

File: src/NzbDrone.Core/Organizer/NamingConfig.cs [hunk 1]
```diff
@@ -8,15 +8,13 @@ public class NamingConfig : ModelBase
         {
             RenameTracks = false,
             ReplaceIllegalCharacters = true,
-            StandardTrackFormat = "{Artist Name} - {Album Title} - {track:00} - {Track Title}",
-            ArtistFolderFormat = "{Artist Name}",
-            AlbumFolderFormat = "{Album Title} ({Release Year})"
+            StandardTrackFormat = "{Album Title} [{Release Year}]/Disc {medium:00}/{Artist Name} - {Album Title} - {track:00} - {Track Title}",
```
Comment (Qstick) [inline src/NzbDrone.Core/Organizer/NamingConfig.cs]: Should this be our default?

**Comment by Qstick** [comment]
Yea I've been thinking about how to do that best. We can build it straight into the logic or we can come up with some sort of string template if then that would allow the user to do it on any naming variable

**Comment by gismo2004** [comment]
something like in C# ? 

`{Album Title} [{Release Year}]({max(medium)} > 1 ? /Disc {medium:00} : "what do use for a blank?")/{Artist Name} - {Album Title} - {track:00} - {Track Title}`

But this would make things very complicated and the standard user will not benefit much of it *(guessing)*

But if it is build-in logic, how to handle `/Disc {medium:00}` vs `/CD {medium:00}`?
Would also require a checkbox to enable/disable it right?

**Comment by Qstick** [comment]
Yea that may get a bit complicated, looking at possibly doing something similar to beets or picard

**Comment by pdf** [comment]
Yeah, I think the frontend probably needs to expose some conditionals, similar to Picard/beets, to handle at least optional components.  Things can get somewhat out of hand though, here's my Picard pattern:

```
$if($eq(%compilation%,1),
$noop(Various Artist albums)
Unfiled/$if2(%albumartist%,%artist%)/%album%$if2($if(%originaldate%, [$left(%originaldate%,4)]),$if(%date%, [$left(%date%,4)]))$if(%catalognumber%, [%catalognumber%]) [$upper(%_extension%)$if($eq($rreplace(%media%,^.*?\([^ ]+\)\$,\\1),Vinyl),-Vinyl)$if($eq(%media%,Digital Media),-Web)$if($eq(%media%,Cassette),-Tape)$if($gt(%_channels%,2),-$sub(%_channels%,1).1)$if($gt(%_sample_rate%,44100),-%_sample_rate%)$if($gt(%_bits_per_sample%,16),-%bits_per_sample%b)]/$if($gt(%totaldiscs%,1),/$num(%discnumber%,2)$if(%discsubtitle%,. %discsubtitle%)/)$num(%tracknumber%,2). %artist% - %title%,
$noop(Single Artist Albums)
$if2(%genre%,Unfiled)/$if2(%albumartist%,%artist%)/%album%$if2($if(%originaldate%, [$left(%originaldate%,4)]),$if(%date%, [$left(%date%,4)]))$if(%catalognumber%, [%catalognumber%]) [$upper(%_extension%)$if($eq($rreplace(%media%,^.*?\([^ ]+\)\$,\\1),Vinyl),-Vinyl)$if($eq(%media%,Digital Media),-Web)$if($eq(%media%,Cassette),-Tape)$if($gt(%_channels%,2),-$sub(%_channels%,1).1)$if($gt(%_sample_rate%,44100),-%_sample_rate%)$if($gt(%_bits_per_sample%,16),-%bits_per_sample%b)]/$if($gt(%totaldiscs%,1),/$num(%discnumber%,2)$if(%discsubtitle%,. %discsubtitle%)/)$num(%tracknumber%,2). %title%)
```

**Comment by pdf** [comment]
@Qstick can I ask what the plan is for this? Is there a new issue to be tracked? This is the feature that's been stopping me from using Lidarr.

**Comment by Qstick** [comment]
Still plans to do this, this branch is just extremely old. We want to get the various artist work in before starting back on this

**Pull request closed without merging**